### PR TITLE
5th PR of Python/GT4Py implementation (Shallow Water Model Tangent Linear and Adjoint Models)

### PR DESCRIPTION
**Pull request by robgpita**
## Proposed changes

This is the fifth of ~5 PRs that will introduce a Python/GT4Py implementation of the Shallow Water code.

This PR introduces: 

- The Python implementation of the Shallow Water Tangent Linear and Adjoint Models.
- Add Pytests for both of the models.
- Modify shallow_water_state to include get_u(), get_v(), get_h() methods.

Tested on laptop with 1,2,4, & 9 processes:
`mpiexec -n 9 --oversubscribe python -m pytest --with-mpi -s --disable-pytest-warnings`

And for non-mpi tests:
`mpiexec -n 1 python -m pytest`

Waiting on merge of PR #14 into develop. 

## Types of changes

- [ ] Bugfix (Change which fixes an issue)
- [x] New feature (Change which adds functionality)
- [x] Enhancement to existing functionality (non-breaking change which improves existing code or documentation)
- [ ] Breaking change (fix or feature that changes results and requires updates to test baselines)
- [ ] Documentation Update

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/NOAA-GSL/SENA-shallow-water/blob/develop/CONTRIBUTING.md) document
- [x] I have ensured all my changes contribute to a common purpose
- [x] I have verified my change does not add debug code
- [x] I have verified my change does not add code that is commented out
- [x] I have verified unit tests pass locally with my changes
- [x] I have verified my changes adhere to style guidelines
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] I have added necessary documentation (if appropriate)

## Reviewers' checklist

The following is a checklist for reviewers to consider when reviewing this pull request.

#### Implementation

- [ ] Does this code do what it is supposed to do?
- [ ] Can the code be simplified?
- [ ] Does the code add unwanted compile-time or run-time dependencies?
- [ ] Is the code at the correct abstraction level?
- [ ] Is the code modular enough?
- [ ] Does similar functionality already exist in the codebase? If so, why wasn't it used?
- [ ] Can you think of a different solution that is substantially better in terms of code maintenance, readability, or performance?

#### Logic Errors and Bugs

- [ ] Can you think of a use case in which the code does not behave properly?
- [ ] Can you think of any inputs or external events that could break the code?

#### Error handling and logging

- [ ] Is error handling done correctly?
- [ ] Should any logging or debuggging be added or removed?
- [ ] Are error messages user friendly?

#### Usability

- [ ] Is the API intuitive to use?
- [ ] Is the API well documented?

#### Testing and testability

- [ ] Do the tests pass?
- [ ] Is the code testable?
- [ ] Do the tests reasonably cover the code change?
- [ ] Are there additional edge cases or inputs that should be tested?

#### Performance

- [ ] Do you think the code change will negatively impact performance?
- [ ] Do you see any potential to improve the performance of the code?

#### Dependencies

- [ ] Does this change require updates to documenation, configuration?  If so, what that done?
- [ ] Does this change affect backward compatibility?

#### Readability

- [ ] Is the code easy to understand?
- [ ] Can the readability be improved with smaller routines?
- [ ] Can the readability be improved with different subroutine/variable names?
- [ ] Is the data flow understandable?
- [ ] Could comments in the code be improved to add clarity?
- [ ] Is there any commented out code?
- [ ] Does the code conform to the style guide (Fortran style compliance cannot be checked automatically)

### CODE REVIEW

File: test/test_input/test_shallow_water_config.yml [hunk 1]
```diff
@@ -11,6 +11,12 @@ model:
   b0: 4.0
   h0: 5.0
 
+runtime:
```
Comment (christopherwharrop-noaa) [inline test/test_input/test_shallow_water_config.yml]: This section is never used.  Doesn't hurt anything to be here, but this YAML doesn't contain valid settings for a run.  It's only used for testing config constructors.

**Review by christopherwharrop-noaa — COMMENTED** [review]
Once other PRs are approved/merged, and this one is brought up to date and passes tests I will approve it.  The code looks great. Thanks for adding the simplifying get_u/v/h methods.

File: src/python/shallow_water_visualization.py [hunk 2]
```diff
@@ -33,6 +33,62 @@ def inverse(self, value):
 
         return np.interp(value, x, y, left=-np.inf, right=np.inf)
 
+<<<<<<< HEAD
```
Comment (christopherwharrop-noaa) [inline src/python/shallow_water_visualization.py]: Not sure if you're still working on resolving conflicts, but your last commit may have included incomplete merge.

File: src/python/shallow_water_visualization.py [hunk 3]
```diff
@@ -33,6 +33,62 @@ def inverse(self, value):
...
+## Plot animation
+X, Y = np.meshgrid(np.linspace(0, g.xmax, g.npx), np.linspace(0, g.ymax, g.npy))
+fig, axs = plt.subplots()
+norm = ColormapNormalize(vmin=4990, vmax=5030, v1=4995, v2=5005)
+pc = axs.pcolormesh(X, Y, s.h.data, cmap='nipy_spectral', norm=norm)
+cb = fig.colorbar(pc, ax=axs, extend='both')
+cb.set_ticks([4990, 4995, 4996, 4997, 4998, 4999, 5000, 5001, 5002, 5003, 5004, 5005, 5030])
+anim = animation.FuncAnimation(fig, animate, interval=125, frames=runtime['run_steps'])
+anim.save('../../test/test_output/test.gif')
+=======
```
Comment (christopherwharrop-noaa) [inline src/python/shallow_water_visualization.py]: See above

File: src/python/shallow_water_visualization.py [hunk 4]
```diff
@@ -113,3 +169,4 @@ def main():
 if __name__ == "__main__":
     main()
 
+>>>>>>> develop
```
Comment (christopherwharrop-noaa) [inline src/python/shallow_water_visualization.py]: And again.

File: src/python/shallow_water_visualization.py (hunk 2 above)
Comment (robgpita) [reply src/python/shallow_water_visualization.py]: Yep, pushed the commit a little early. Still working on cleaning this up.

**Comment by christopherwharrop-noaa** [comment]
@robgonzalezpita  - This looks ready to convert to a real PR to me.  The tests for the TL and ADJ all look great!

**Review by christopherwharrop-noaa — APPROVED** [review]
Great addition.  Thank you for the improvements.
